File: app/controllers/lexical_diversity.rb
class LexicalDiversity

  def lex_d(text, mtld_ttr_threshold=0.72, hdd_sample_size=40.0)
    # puts mtld(text, mtld_ttr_threshold)
    # puts hdd(text, hdd_sample_size)
    # puts yules_i(text)
    (mtld(text, mtld_ttr_threshold) + hdd(text, hdd_sample_size) + yules_i(text)) / 3
  end

  def mtld(text, ttr_threshold=0.72)
    text_array = clean_text(text)
    return 0 if text_array.empty?

    val1 = mtld_eval(text_array, ttr_threshold)
    val2 = mtld_eval(text_array.reverse, ttr_threshold)
    mtld_scale((val1 + val2) / 2.0)
  end

  def mtld_eval(text_array, ttr_threshold)
    current_ttr = 1.0
    current_types = 0.0
    current_tokens = 0.0
    current_words = []
    factors = 0.0

    text_array.each do |element|
      current_tokens += 1
      unless current_words.include?(element)
        current_types += 1
        current_words << element
      end

      current_ttr = current_types / current_tokens

      if current_ttr < ttr_threshold
        factors += 1
        current_ttr = 0.0
        current_types = 0.0
        current_tokens = 0.0
        current_words = []
      end
    end

    excess = 1.0 - current_ttr
    excess_val = 1.0 - ttr_threshold
    factors += excess / excess_val

    text_array.size / factors
  end

  def mtld_scale(mtld)
    ((mtld - 99.284) * 0.5554 + 100)
  end

  def hdd(text, sample_size=40.0)
    token_array = clean_text(text)
    return 0 if token_array.empty?
    hdd_value = 0.0

    type_array = create_type_array(token_array)

    type_array.each do |element|
      contribution = 1.0 - hypergeometric(token_array.size, sample_size, token_array.count(element), 0.0)
      contribution = contribution / sample_size
      hdd_value += contribution
    end

    hdd_scale(hdd_value)
  end

  def hypergeometric(population, sample, pop_successes, samp_successes)
    (combination(pop_successes, samp_successes) * combination(population - pop_successes, sample - samp_successes)) / combination(population, sample)
  end

  def combination(n, k)
    n_minus_k = n - k
    i = n
    numerator = 1
    while i > n_minus_k && i > 0 do
      numerator *= i
      i -= 1
    end
    numerator / factorial(k)
  end

  def factorial(n)
    if n <= 1
      1
    else
      n * factorial(n - 1)
    end
  end

  def hdd_scale(hdd)
    ((hdd - 0.854) * 592.1052 + 100)
  end

  def yules_i(text)
    token_array = clean_text(text)
    return 0 if token_array.empty?

    type_array = create_type_array(token_array)

    m1 = token_array.size
    m2 = 0.0
    freq_array = Array.new(type_array.size / 2.0, 0.0)

    type_array.each do |element|
      return 0 if token_array.count(element) >= freq_array.size
      freq_array[token_array.count(element)] += 1.0
    end

    freq_array.each_with_index do |element, index|
      m2 += (element * (index ** 2))
    end
    return 0 if (m2 - m1) == 0
    yules_scale((m1 * m1) / (m2 - m1))
  end

  def yules_scale(yules)
    ((yules - 100.793) * 0.6818 + 100)
  end

  def clean_text(text)
    new_text = text.gsub(/<(.|\n)*?>/, ' ')
    new_text = new_text.gsub(/&nbsp;/, ' ')
    new_text = new_text.gsub(/[^a-z0-9 ]/i, '')
    new_text = new_text.downcase
    new_text.split
  end

  def create_type_array(token_array)
    type_array = []
    token_array.each do |element|
      unless type_array.include?(element)
        type_array << element
      end
    end
    type_array
  end
end

File: app/controllers/lexical_diversity.js
function () {

  var lex_d = function (args) {
    var text = args["text"];
    var ttr_threshold = "mtld_ttr_threshold" in args ? args["mtld_ttr_threshold"] : 0.72;
    var sample_size = "hdd_sample_size" in args ? args["hdd_sample_size"] : 40.0;

    return (mtld({'text': text, 'ttr_threshold': ttr_threshold}) + 
            hdd({'text': text, 'sample_size': sample_size}) + 
            yules_i(text)) / 3;
  }

  var mtld = function (args) {
    var text = args["text"];
    var ttr_threshold = "ttr_threshold" in args ? args["ttr_threshold"] : 0.72;
    var text_array = clean_text(text);
    if (text_array === null) { return mtld_scale(0); }

    var val1 = mtld_eval(text_array, ttr_threshold);
    var val2 = mtld_eval(text_array.reverse(), ttr_threshold);
    
    return mtld_scale((val1 + val2) / 2);
  }

  var mtld_eval = function (text_array, ttr_threshold) {
    var current_ttr = 1.0,
    current_types = 0.0,
    current_tokens = 0.0,
    current_words = [],
    factors = 0.0;

    text_array.forEach(function(word) {
      current_tokens++;
      if (current_words.indexOf(word) == -1) {
        current_types++;
        current_words.push(word);
      }

      if (current_tokens === 0.0) { return 0; }
      current_ttr = current_types / current_tokens;

      if (current_ttr < ttr_threshold) {
        factors++;
        current_ttr = 0.0;
        current_types = 0.0;
        current_tokens = 0.0;
        current_words = [];
      }
    });

    var excess = 1.0 - current_ttr;
    var excess_val = 1.0 - ttr_threshold;
    if (excess_val === 0.0) { return 0; }
    factors += excess / excess_val;

    if (factors === 0.0) { return 0; }
    return text_array.length / factors;
  }

  var mtld_scale = function (mtld) {
    return ((mtld - 99.284) * 0.5554 + 100);
  }

  var hdd = function (args) {
    var text = args["text"];
    var sample_size = "sample_size" in args ? args["sample_size"] : 40.0;
    var token_arrray = clean_text(text);
    if (token_arrray === null) { return hdd_scale(0); }
    
    var hdd_value = 0.0;

    var type_array = create_type_array(token_arrray);

    for (var i = 0; i < type_array.length; i++) {
      var count = 0;
      for (var j = 0; j < token_arrray.length; j++) {
        if (type_array[i] == token_arrray[j]) {
          count++;
        }
      }
      var contribution = 1.0 - hypergeometric(token_arrray.length, sample_size, count, 0.0);
      contribution /= sample_size;
      hdd_value += contribution;
    }
    return hdd_scale(hdd_value);
  }

  var hypergeometric = function (population, sample, pop_successes, samp_successes) {
    var mk = combination(pop_successes, samp_successes);
    var Nmnk = combination(population - pop_successes, sample - samp_successes);
    return (mk * Nmnk) / combination(population, sample);
  }

  var combination = function (n, k) {
    var i = n;
    var numerator = 1;
    while (i > (n - k) && i > 0) {
      numerator *= i;
      i -= 1;
    }
    return numerator / factorial(k);
  }

  var factorial = function (n) {
    if (n <= 1) {
      return 1;
    }
    return (n * factorial(n - 1));
  }

  var hdd_scale = function(hdd) { 
    return ((hdd - 0.854) * 592.1052 + 100); 
  }

  var yules_i = function (text) {
    var token_arrray = clean_text(text);
    if (token_arrray === null) { return yules_scale(0); }
    var type_array = create_type_array(token_arrray);

    var m1 = token_arrray.length;
    var m2 = 0.0;
    var new_arr_length = parseInt(type_array.length/2);
    var freq_array = Array.apply(null, new Array(new_arr_length)).map(Number.prototype.valueOf,0);

    for (var i = 0; i < type_array.length; i++) {
      var count = 0;
      for (var j = 0; j < token_arrray.length; j++) {
        if (type_array[i] == token_arrray[j]) {
          count++;
        }
      }
      if (count >= freq_array.length) { return 0; }
      freq_array[count] += 1;
    }

    for (var i = 0; i < freq_array.length; i++) {
      m2 += (freq_array[i] * (i * i));
    }
    if ((m2 - m1) === 0) { return yules_scale(0); }
    return yules_scale((m1 * m1) / (m2 - m1));
  }

  var yules_scale = function(yules) {
    return ((yules - 100.793) * 0.6818 + 100)
  }

  var create_type_array = function (token_arrray) {
    var type_array = [];
    for (var i = 0; i < token_arrray.length; i++) {
      if (type_array.indexOf(token_arrray[i]) == -1) {
        type_array.push(token_arrray[i]);
      }
    }
    return type_array;
  }

  var clean_text = function (text) {
    var html_strip = /<(.|\n)*?>/;
    var new_text = text;
    while (html_strip.test(new_text)) {
      new_text = new_text.replace(html_strip, ' ');
    }
    new_text = new_text.replace(/&nbsp;/g, ' ');
    new_text = new_text.replace(/\\n/g, ' ');
    new_text = new_text.replace(/[^\w\s\d]/gi, '');
    return new_text.toLowerCase().match(/[a-z0-9]+/gi);
  }

  if (this.revs && this.revs.length > 0){
    var last = this.revs.length - 1 ;
    if (this.revs[last].length > 1 && this.revs[last][1].length > 0){
      var doc = this.revs[last][1];
      //emit(this._id, mtld({'text': doc}));
      //emit(this._id, hdd({'text': doc}));
      //emit(this._id, yules_i(doc));
      emit(this._id, lex_d({'text': doc}));
    }
  }
}
